onmt/trainer.py

```diff
@@ -447,6 +447,11 @@ def validate(self, valid_iter, moving_average=None, task=None):
         # Set model back to training mode.
         valid_model.train()
 
+        # the forward hook `has_grad` was triggered, so we manually unset the flags to not fool the optim
```

Comment:
This should work, as long as validation is never performed in the middle of an accumulation (a very reasonable assumption).

However, I think a better solution would be to modify the hook: have it check that `module.training` is set. If unset, then return without doing anything.

Reply:
the efficient thing would be to set `param.has_grad = self.training`, but that's a case of clarity vs. efficiency.